# Лабораторная работа № 4
## Хапчаев Тимур Русланович M32071
### Вариант 4

In [22]:
import numpy as np
import pandas as pd
import scipy.stats as sps
from statsmodels.stats.diagnostic import lilliefors

data = pd.read_csv("/Users/tkhapchaev/Documents/GitHub/mathematical-statistics/lab-4/song_data.csv")
data.head()

song_name  song_popularity  song_duration_ms   
0  Boulevard of Broken Dreams               73            262333  \
1                  In The End               66            216933   
2           Seven Nation Army               76            231733   
3                  By The Way               74            216933   
4           How You Remind Me               56            223826   

   acousticness  danceability  energy  instrumentalness  key  liveness   
0      0.005520         0.496   0.682          0.000029    8    0.0589  \
1      0.010300         0.542   0.853          0.000000    3    0.1080   
2      0.008170         0.737   0.463          0.447000    0    0.2550   
3      0.026400         0.451   0.970          0.003550    0    0.1020   
4      0.000954         0.447   0.766          0.000000   10    0.1130   

   loudness  audio_mode  speechiness    tempo  time_signature  audio_valence  
0    -4.095           1       0.0294  167.060               4          0.474  
1    -6.407           0       0.0498  105.256               4          0.370  
2    -7.828           1       0.0792  123.881               4          0.324  
3    -4.938           1       0.1070  122.444               4          0.198  
4    -5.065           1       0.0313  172.011               4          0.574

In [23]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18835 entries, 0 to 18834
Data columns (total 15 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   song_name         18835 non-null  object 
 1   song_popularity   18835 non-null  int64  
 2   song_duration_ms  18835 non-null  int64  
 3   acousticness      18835 non-null  float64
 4   danceability      18835 non-null  float64
 5   energy            18835 non-null  float64
 6   instrumentalness  18835 non-null  float64
 7   key               18835 non-null  int64  
 8   liveness          18835 non-null  float64
 9   loudness          18835 non-null  float64
 10  audio_mode        18835 non-null  int64  
 11  speechiness       18835 non-null  float64
 12  tempo             18835 non-null  float64
 13  time_signature    18835 non-null  int64  
 14  audio_valence     18835 non-null  float64
dtypes: float64(9), int64(5), object(1)
memory usage: 2.2+ MB


### Задание 1

Пусть $X$ - случайная величина, популярность песни. Тогда нулевая гипотеза имеет вид:<br>
<br>
$H_0: X$ имеет нормальное распределение, <br>
<br>
альтернативная гипотеза имеет вид:<br>
<br>
$H_1: X$ имеет распределение, отличное от нормального.<br>
<br>
Для проверки данного утверждения воспользуемся двумя критериями согласия - критерием согласия Лиллиефорса, и критерием согласия Шапиро-Уилка. Далее везде используем уровень значимости $\alpha = 0.05$.

Статистика критерия Лиллиефорса имеет вид $D_n = \underset{x \in \mathbb{R}}{\sup} |\widehat{F_n}(x) - F_0(x)|$, где $\widehat{F_n}(x)$ - эмпирическая функция распределения выборки в точке $x$, $F_0(x)$ - теоретическая функция распределения случайной величины, согласие с законом распределения которой тестируется, причем неизвестные параметры этого теоретического распределения оцениваются по имеющейся выборке (в нашем случае $F_0(x)$ - это функция нормального распределения, для которого математическое ожидание и дисперсия оцениваются как выборочное среднее и выборочная несмещенная дисперсия по имеющейся выборке популярности песни соответственно), $n$ - объем имеющейся выборки.
Стоит отметить, что тест Лиллиефорса - это, фактически, расширение критерия согласия Колмогорова на случай неизвестных параметров тестируемого распределения (применение этих критериев корректно в случае непрерывных функций распределения $F(x)$ и $F_0(x)$, в нашем случае мы предполагаем непрерывность величины $X$ в силу специфики данных, и непрерывность функции нормального распределения $F_0(x)$ очевидна).
Нулевая гипотеза отвергается на уровне значимости $\alpha$, если значение статистики $D_n$, посчитанное по имеющей выборке, превышает критическое значение распределения Лиллиефорса (при истинной нулевой гипотезе), соответствующее уровню значимости $\alpha$. Данные критические значения для разных объемов выборки приведены в таблицах. В нашем случае для 5% уровня значимости критическое значение определяется по формуле $0.886/\sqrt{n}$, и равно $0.886/\sqrt{18835} \approx 0.006$.<br>
<br>
Посчитаем значение статистики и p-value теста. Если p-value получится меньше, либо равно 0.05, то отвергаем нулевую гипотезу (так как это эквивалентно тому, что значение статистики критерия по выборке превысило критическое значение).

In [24]:
0.886 / np.sqrt(18835)

0.0064558125177343805

In [25]:
test_value = round(lilliefors(x = data["song_popularity"], dist = "norm")[0], ndigits = 3)
p_value = round(lilliefors(x = data["song_popularity"], dist = "norm")[1], ndigits = 3)

print(f'Значение тестовой статистики и p-value теста Лиллиефорса составили, соответственно: {test_value} и {p_value}.')

Значение тестовой статистики и p-value теста Лиллиефорса составили, соответственно: 0.062 и 0.001.


In [26]:
mean = (data["song_popularity"] - np.mean(data["song_popularity"]))
var = np.sqrt(np.var(data["song_popularity"], ddof = 1))

eq = sps.kstest(mean / var, "norm")[0] == lilliefors(data["song_popularity"], dist = "norm")[0]
eq

True

Значение статистики совпадает с полученной выше для критерия Колмогорова, где данные отскалированы на выборочное среднее и несмещенную дисперсию, посчитанные по выборке.

Таким образом, имеем значение статистики по выборке в 0.062, что больше критического значения 0.006 (аналогично, p-value 0.001 меньше, чем 0.05), из чего следует отвержение нулевой гипотезы на уровне значимости 0.05. Другими словами, с максимальной вероятностью ошибки первого рода в 0.05 можно утверждать, что популярность песни имеет распределение, отличное от нормального.

Тестовая статистика критерия Шапиро-Уилка имеет вид $\dfrac{(\sum_{i=1}^{n} a_i x_{(i)})^2}{\sum_{i=1}^{n} (x_i - \overline{x})^2}$, где $x_{(i)}$ - i-я порядковая статистика по выборке (i-е по порядку значение популярности песни при упорядочении выборки по возрастанию), $\overline{x} = \dfrac{\sum_{i=1}^{n} x_i}{n}$ - выборочное среднее, $a_i$ - коэффициенты, определяемые по специальным формулам и таблицам. <br>
<br>

Нулевая гипотеза отвергается на уровне значимости $\alpha$, если значение статистики $W$, посчитанное по имеющей выборке, превышает некоторое критическое значение (посчитанное при истинной нулевой гипотезе), соответствующее уровню значимости $\alpha$.
<br>
Посчитаем теперь значение статистики и p-value теста:

In [27]:
test_value = round(sps.shapiro(data["song_popularity"])[0], ndigits = 3)
p_value = round(sps.shapiro(data["song_popularity"])[1], ndigits = 3)

print(f'Значение тестовой статистики и p-value теста Шапиро-Уилка составили, соответственно: {test_value} и {p_value}.')

Значение тестовой статистики и p-value теста Шапиро-Уилка составили, соответственно: 0.973 и 0.0.


/Library/Python/3.9/site-packages/scipy/stats/_morestats.py:1816: UserWarning: p-value may not be accurate for N > 5000.
  warnings.warn("p-value may not be accurate for N > 5000.")


In [28]:
sps.shapiro(data["song_popularity"])

ShapiroResult(statistic=0.9732725024223328, pvalue=0.0)

Таким образом, имеем значение p-value около 0, что меньше, чем 0.05, откуда с максимальной вероятностью ошибки первого рода в 0.05 можем отвергнуть нулевую гипотезу о нормальном распределении популярности песни.

Имеем отвержение гипотезы нормальности (на уровне значимости 0.05) популярности песни в обоих случаях, таким образом аппроксимировать популярность песни нормальным законом не следует.

### Задание 2

In [29]:
data.head()

song_name  song_popularity  song_duration_ms   
0  Boulevard of Broken Dreams               73            262333  \
1                  In The End               66            216933   
2           Seven Nation Army               76            231733   
3                  By The Way               74            216933   
4           How You Remind Me               56            223826   

   acousticness  danceability  energy  instrumentalness  key  liveness   
0      0.005520         0.496   0.682          0.000029    8    0.0589  \
1      0.010300         0.542   0.853          0.000000    3    0.1080   
2      0.008170         0.737   0.463          0.447000    0    0.2550   
3      0.026400         0.451   0.970          0.003550    0    0.1020   
4      0.000954         0.447   0.766          0.000000   10    0.1130   

   loudness  audio_mode  speechiness    tempo  time_signature  audio_valence  
0    -4.095           1       0.0294  167.060               4          0.474  
1    -6.407           0       0.0498  105.256               4          0.370  
2    -7.828           1       0.0792  123.881               4          0.324  
3    -4.938           1       0.1070  122.444               4          0.198  
4    -5.065           1       0.0313  172.011               4          0.574

In [30]:
border_duration = round(np.quantile(data["song_duration_ms"], 0.6))

In [31]:
border_duration

221512

Задачу тестирования значимости отличий рейтинга популярности "длинных" и "коротких" песен можно свести к задаче тестирования гипотезы о том, что выборки популярности длинных и коротких песен взяты из одной и той же генеральной совокупности, то есть гипотезы однородности. Итак, пусть $X$ - случайная величина - популярность коротких песен, $Y$ - случайная величина - популярность длинных песен. Сделаем разбиение на длинные и короткие песни следующим образом: те песни, у которых продолжительность не превышает выборочный квантиль уровня 0.6 по имеющейся выборке продолжительности песен (округленный до целого числа) будем считать короткими, остальные - длинными. Мы проверяем гипотезу:<br>
<br>
$H_0: F(x) = G(x)$ для всех $x \in \mathbb R$

$H_1: F(x) \neq G(x)$ для всех $x \in \mathbb R$,<br>
<br>
где $F(x)$ и $G(x)$ - функции распределения случайных величин $X$ и $Y$ соответственно, причем эти функции распределения непрерывны (сделаем такое предположение исходя из данных, которые не дискретны). <br>
<br>
Первый способ тестирования - ***критерий однородности Колмогорова***.<br>
<br>
Если $H_0$ верна, то $F(x) = G(x) = F_0(X)$

 $F^*_n$ эмпирическая ф-я распределения, построенная по выборке $X_{[n]}$

 $G^*_m$ эмпирическая ф-я распределения, построенная по выборке $Y_{[m]}$<br>
 <br>
 Статистика критерия имеет вид $D_{m,n} = \overset{}{\underset{x \in \mathbb R}{sup}} | G^*_m(x, Y_{[m]}) - F^*_n(x, X_{[n]})|$<br>
 <br>
 Если $F_0(x) = F(x) = G(x)$ непрерывна, тогда

$P(\sqrt{\frac{mn}{m + n}} D_{m, n} < z) \xrightarrow[n, m \rightarrow \infty]{} K(z) = \sum \limits_{j = - \infty}^{\infty} (-1)^j e^{-2j^2z^2}$, где

$K(z)$ - ф-я распред. Колмогорова.<br>
<br>
Критическая область:<br>
<br>
$V_k = (d_{1 - \alpha}, + \infty)$, где

$d_{1 - \alpha}$ квантиль Колмогорова.

$\sqrt{\frac{mn}{m + n}} D_{m, n} \in V_k \Rightarrow H_0$ отвергается.<br>
<br>
Рассчитаем теперь p-value теста для нашей задачи, если оно меньше, либо равно 0.05 (что эквивалентно превышению значением статистики критерия по выборке критического значения), то отвергаем нулевую гипотезу на 5% уровне значимости.

In [32]:
popularity_long = data[data["song_duration_ms"] > border_duration]["song_popularity"]
popularity_short = data[data["song_duration_ms"] <= border_duration]["song_popularity"]

eq = len(popularity_short) + len(popularity_long) == len(data["song_popularity"])
eq

True

In [33]:
popularity_long[:5]

0    73
2    76
4    56
5    80
8    80
Name: song_popularity, dtype: int64

In [34]:
popularity_short[:5]

1    66
3    74
6    81
7    76
9    81
Name: song_popularity, dtype: int64

In [35]:
sps.ks_2samp(data1 = popularity_short, data2 = popularity_long)[0]

0.03172285638439076

In [36]:
test_value = round(sps.ks_2samp(data1 = popularity_short, data2 = popularity_long)[0], ndigits = 3)
p_value = round(sps.ks_2samp(data1 = popularity_short, data2 = popularity_long)[1], ndigits = 3)

print(f'Значение тестовой статистики и p-value критерия однородности Колмогорова составили, соответственно: {test_value} и {p_value}.')

Значение тестовой статистики и p-value критерия однородности Колмогорова составили, соответственно: 0.032 и 0.0.


Таким образом, имеем p-value около 0, что меньше 0.05, то есть отвергаем нулевую гипотезу, и можем с максимальной вероятностью ошибки первого рода в 0.05 утверждать, что распределение популярности песни отличается в зависимости от продолжительности.

Второй способ тестирования - критерий однородности ***Андерсона-Дарлинга***.<br>
<br>
Статистика критерия имеет вид (для непрерывных функций распределения $F(x)$ и $G(x)$ в нулевой гипотезе выше):<br>
<br>
$A_2 = \dfrac{1}{mn}\sum_{i=1}^{n+m-1} \dfrac{(M_i (m+n) - mi)^2}{i(m+n-i)}$,<br>
<br>
где $M_i$ - число элементов первой выборки, меньших или равных i-му элементу вариационного ряда объединенной выборки.<br>
<br>
Нулевая гипотеза отвергается на уровне значимости 0.05 при превышении статистикой критерия соответствующего критического значения.<br>
<br>
Посчитаем, как и ранее, p-value теста.

In [37]:
sps.anderson_ksamp(samples = [popularity_short, popularity_long])

/var/folders/l_/grbqmmln6bq84j3d3dt_jvqc0000gn/T/ipykernel_67496/2408101600.py:1: UserWarning: p-value floored: true value smaller than 0.001
  sps.anderson_ksamp(samples = [popularity_short, popularity_long])


Anderson_ksampResult(statistic=7.588853157969736, critical_values=array([0.325, 1.226, 1.961, 2.718, 3.752, 4.592, 6.546]), pvalue=0.001)

In [38]:
test_value = round(sps.anderson_ksamp(samples = [popularity_short, popularity_long])[0], ndigits = 3)
p_value = round(sps.anderson_ksamp(samples = [popularity_short, popularity_long])[2], ndigits = 3)

print(f'Значение тестовой статистики и p-value критерия однородности Андерсона-Дарлинга составили, соответственно: {test_value} и {p_value}')

Значение тестовой статистики и p-value критерия однородности Андерсона-Дарлинга составили, соответственно: 7.589 и 0.001


/var/folders/l_/grbqmmln6bq84j3d3dt_jvqc0000gn/T/ipykernel_67496/2070441604.py:1: UserWarning: p-value floored: true value smaller than 0.001
  test_value = round(sps.anderson_ksamp(samples = [popularity_short, popularity_long])[0], ndigits = 3)
/var/folders/l_/grbqmmln6bq84j3d3dt_jvqc0000gn/T/ipykernel_67496/2070441604.py:2: UserWarning: p-value floored: true value smaller than 0.001
  p_value = round(sps.anderson_ksamp(samples = [popularity_short, popularity_long])[2], ndigits = 3)


Таким образом, значение статистики критерия составило 7.589, что больше, чем критическое значение для 5% уровня значимости, которое равно 1.961 (аналогично, p-value теста составило величину, меньше, чем 0.001 < 0.05), из чего следует отвержение нулевой гипотезы на уровне значимости 0.05, поэтому можем с максимальной вероятностью ошибки первого рода в 0.05 утверждать, что выборки популярности длинных и коротких песен получены из разных генеральных совокупностей, то есть имеют разные распределения.

In [39]:
len(popularity_short), len(popularity_long)

(11301, 7534)

In [40]:
np.mean(popularity_short), np.mean(popularity_long)

(53.26245465003097, 52.586010087602865)

In [41]:
np.var(popularity_short), np.var(popularity_long)

(483.6359224412007, 473.85210584858413)

В данном случае заключения двух критериев совпадают, и на уровне значимости 5% можем утверждать, что есть принципиальные отличия в популярности длинных и коротких песен.

### Задание 3

In [42]:
data.head()

song_name  song_popularity  song_duration_ms   
0  Boulevard of Broken Dreams               73            262333  \
1                  In The End               66            216933   
2           Seven Nation Army               76            231733   
3                  By The Way               74            216933   
4           How You Remind Me               56            223826   

   acousticness  danceability  energy  instrumentalness  key  liveness   
0      0.005520         0.496   0.682          0.000029    8    0.0589  \
1      0.010300         0.542   0.853          0.000000    3    0.1080   
2      0.008170         0.737   0.463          0.447000    0    0.2550   
3      0.026400         0.451   0.970          0.003550    0    0.1020   
4      0.000954         0.447   0.766          0.000000   10    0.1130   

   loudness  audio_mode  speechiness    tempo  time_signature  audio_valence  
0    -4.095           1       0.0294  167.060               4          0.474  
1    -6.407           0       0.0498  105.256               4          0.370  
2    -7.828           1       0.0792  123.881               4          0.324  
3    -4.938           1       0.1070  122.444               4          0.198  
4    -5.065           1       0.0313  172.011               4          0.574

Рассмотрим две случайные величины: $Z$ - популярность песни, и $D$ - ее продолжительность. Зависимость популярности песни от продолжительности по определению означает, что при изменении значения $D$ меняется распределение $Z$, то есть условная функция распределения $F(Z | D)$ зависит от значений $D$. Введем теперь случайную величину $D_{\text{binar}} = 1, \ \text{при} \ D > 221512, \ \text{и} \ D_{\text{binar}} = 0 \ \text{иначе}$. Очевидно, это измеримая функция от случайной величины $D$. Заметим теперь, что в задаче 2 мы протестировали фактически зависимость случайных величин $Z$ и $D_{\text{binar}}$, так как проверка отличия распределения $Z$ в зависимости от того, длинная песня ($D_{\text{binar}} = 1$) или короткая ($D_{\text{binar}} = 0$), эквивалентна фактически проверке утверждения, что распределение $Z$ зависит от значений $D_{\text{binar}}$. Мы получили отвержение нулевой гипотезы во 2 задаче, и заключили, что распределение $Z$ принципиально отличается в случаях $D_{\text{binar}} = 1$ и $D_{\text{binar}} = 0$. Другими словами, случайные величины $Z$ и $D_{\text{binar}}$ зависимы по определению. Так как $D_{\text{binar}}$ - это функция от величины $D$, то при изменении значений $D$ меняются значения и $D_{\text{binar}}$, а значит меняется и распределение $Z$, то есть величины $Z$ и $D$ тоже зависимы (по вероятностному определению) - если песня "короткая", то популярность будет иметь другое распределение, чем если бы песня была "длинной". Формально, тестирование зависимости здесь эквивалентно тестированию утверждений в задаче 2, то есть гипотеза о независимости эквивалентна конкретной гипотезе об однородности (хотя стоит отметить, что если бы мы не отвергли гипотезу об однородности в задаче 2, это еще не означало бы независимость. Но если мы хотя бы для одного порога отвергаем гипотезу об однородности, то это уже означает наличие зависимости величин $Z$ и $D$ по определению).